core/primitives-core/src/account.rs

```diff
@@ -104,39 +177,194 @@ impl Account {
+                    None => {
+                        return Err(serde::de::Error::custom("Missing `version` field"));
+                    }
+                };
+
+                #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
+                if version < AccountVersion::V2 && nonrefundable > 0 {
+                    return Err(serde::de::Error::custom(
+                        "Non-refundable positive amount exists for account version older than V2",
+                    ));
+                }
+
+                Ok(Account {
+                    amount: account_data.amount,
+                    locked: account_data.locked,
+                    code_hash: account_data.code_hash,
+                    storage_usage: account_data.storage_usage,
+                    nonrefundable,
+                    version,
+                })
+            }
+            None => Ok(Account {
+                amount: account_data.amount,
+                locked: account_data.locked,
+                code_hash: account_data.code_hash,
+                storage_usage: account_data.storage_usage,
+                nonrefundable: 0,
+                version: AccountVersion::V1,
+            }),
+        }
+    }
+}
+
 impl BorshDeserialize for Account {
     fn deserialize_reader<R: io::Read>(rd: &mut R) -> io::Result<Self> {
-        // This should only ever happen if we have pre-transition account serialized in state
-        // See test_account_size
-        let deserialized_account = LegacyAccount::deserialize_reader(rd)?;
-        Ok(Account {
-            amount: deserialized_account.amount,
-            locked: deserialized_account.locked,
-            code_hash: deserialized_account.code_hash,
-            storage_usage: deserialized_account.storage_usage,
-            version: AccountVersion::V1,
-        })
+        // The first value of all Account serialization formats is a u128,
+        // either a sentinel or a balance.
+        let sentinel_or_amount = u128::deserialize_reader(rd)?;
+        if sentinel_or_amount == Account::SERIALIZATION_SENTINEL {
+            // Account v2 or newer.
+            let version_byte = u8::deserialize_reader(rd)?;
+            let version = AccountVersion::try_from(version_byte).map_err(|_| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "Error deserializing account: invalid account version {}",
+                        version_byte
+                    ),
+                )
+            })?;
+            #[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
+            if version < AccountVersion::V2 {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("Expected account version 2 or higher, got {:?}", version),
+                ));
+            }
+            let account = AccountV2::deserialize_reader(rd)?;
```

Comment:
Doesn't this allow `SENTINEL | 1 | AccountV2` to be successfully deserialized when `not(protocol_feature_nonrefundable_transfer_nep491)`? My thinking is that when this feature is disabled we should fail deserialization as we encounter the SENTINEL.

This also allows for multiple serialized forms that deserialize to the same thing.

Reply:
Yes, this branch is never entered if `not(protocol_feature_nonrefundable_transfer_nep491)`.
I've added an error return at the beginning of this branch.

The reason why we have
```
#[cfg(feature = "protocol_feature_nonrefundable_transfer_nep491")]
if version < AccountVersion::V2 {
```
is because `AccountVersion::V2` is introduced by `protocol_feature_nonrefundable_transfer_nep491`.

Reply:
Can you please walk me through why you claim this branch is never entered? My comment is operating under an assumption that anybody can handcraft some serialized data or is using fuzzer to generate inputs for this function, and that any reasoning based on what "trusted" serialization code does is invalid.

So my reading is that I can trivially get the code succeed the `sentinel_or_amount` check by starting my hand crafted payload with `0xFFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF` and then there's nothing stopping me from setting the next field for `version_byte` to `1` and finishing with an `AccountV2` structure anyway.

The panic you have added suggests that this function never deals with any data that comes from potentially untrusted external sources, but that seems pretty difficult claim to verify. And it definitely needs to stand, as otherwise this panic here can have pretty severe consequences if a nefarious actor finds a way to execute it. In the worst case the entire network can be brought down for an extended time and do pretty nasty stuff during that downtime.

Reply:
Ah, I see now what you were up to.
Yes, the code potentially allowed to create V1 account using a sentinel.
I've changed it to return an error instead of panicking.
My previous comment was operating under an assumption that this branch is never entered under a `normal` operation in not-nightly and if there is a malicious input then one could set an arbitrary account `amount` anyway.
But yes, we have no guarantee how this code will be used so we must not panic there.